fix(browser): ensure browser state is EXECUTING when tests start

solo-md opened the pull request:
Browser state is EXECUTING when it actually started to execute tests. This state change is triggered by client on actual tests execution start.

Fixes https://github.com/karma-runner/karma/issues/1640

solo-md commented:
I signed it!

johnjbarton commented on lib/browser.js:
```diff
@@ -101,6 +101,8 @@ class Browser {
     this.lastResult.total = info.total
 
+    this.state = EXECUTING
+
```
Changing `this.state` in client callbacks responding to events makes sense: we should to this uniformly unless there is some oddity which prevents us.  Therefore please try to remove the `this.state = EXECUTING` in the client command `execute()`.  The `execute()` command should not set the state; it should be set as you do here, in the start callback.

solo-md replied:
I agree that `execute()` command shouldn't set state to `EXECUTING`, as actually tests are not executing at that moment and before `onStart()`. But we should make a difference between `READY` state of the browser and the state which is after `execute()` and before `onStart()`. We want at least the browser's methods `onKarmaError()` and `onInfo()`during this state. 

I've added additional state called CONFIGURING, which explains the actual state between `execute()` and `onStart()`. Is that OK?

johnjbarton replied:
Yes! This is better. How about one more improvement?  The READY state is IMO misnamed.  After this change the states are:
```js
Browser.STATE_READY = READY
Browser.STATE_CONFIGURING = CONFIGURING
Browser.STATE_EXECUTING = EXECUTING
Browser.STATE_READY_DISCONNECTED = READY_DISCONNECTED
Browser.STATE_EXECUTING_DISCONNECTED = EXECUTING_DISCONNECTED
Browser.STATE_DISCONNECTED = DISCONNECTED
```
But I think the code would be clearer if the initial state was `CONNECTED` rather than `READY`:
```js
Browser.STATE_CONNECTED = CONNECTED
Browser.STATE_CONFIGURING = CONFIGURING
Browser.STATE_EXECUTING = EXECUTING
Browser.STATE_READY_DISCONNECTED = READY_DISCONNECTED
Browser.STATE_EXECUTING_DISCONNECTED = EXECUTING_DISCONNECTED
Browser.STATE_DISCONNECTED = DISCONNECTED
```
That clarifies the relationship with others states like `EXECUTING_DISCONNECTED` and avoids ambiguity about what the browser is "ready-for".

This version of the new state would be documented eg:
```js
// The browser is connected but not yet been commanded to execute tests.
  const CONNECTED = 1
```

solo-md replied:
I don't have any strong opinion here about naming.
State `READY_DISCONNECTED` is actually never used (and sounds confusing with the new states). I'll remove it.

solo-md commented:
I also just came from vacation. Now the feedback-fix loop time should decrease.

johnjbarton commented on lib/browser.js:
```diff
@@ -7,17 +7,20 @@ const logger = require('./logger')
 // The browser is ready to execute tests.
 const READY = 1
 
+// The browser is configuring before tests execution.
```
I think a slightly better description would be
```js
// The browser has been told to execute tests; it is configuring before tests execution.

johnjbarton commented:
> I also just came from vacation.  Now the feedback-fix loop time should decrease.

me too ;-)

solo-md commented:
@johnjbarton I've refactored the state names as you've suggested.
State `READY_DISCONNECTED` is removed (because it was not used at all).
The `browser.isReady()` method was also refactored to `browser.isConnected()`.
